New Attachment Upload UX with Real-Time Previews

kidroca opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Show the attachment instead of the text "Uploading attachment..."

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/9402
PROPOSAL: https://github.com/Expensify/App/issues/9402#issuecomment-2112686001



### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Open a Chat and Send Attachments
1. Open a chat.
2. Upload an image.
   - Verify that a preview of the image appears immediately in the chat.
3. Upload a video.
   - Verify that a preview of the video appears immediately in the chat.
4. Upload a PDF document.
   - Verify that a download link for the PDF appears immediately in the chat.
5. Upload a text file.
   - Verify that a download link for the text file appears immediately in the chat.

#### Known Issue
1. Upload an image and wait for the backend to replace the optimistic action.
   - Observe the image flashing briefly as the optimistic action is replaced by the actual action.
   - Note: This is a known limitation, as agreed upon in [this discussion](https://github.com/Expensify/App/issues/9402#issuecomment-2178981285), to be addressed in a follow-up.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

1. Disconnect from the internet.
2. Upload an image.
   - Verify that a preview of the image appears immediately in the chat and is slightly grayed out, indicating an optimistic action.
3. Reconnect to the internet.
   - Verify that the image preview updates to full opacity once the upload is confirmed by the backend.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as the Tests section above

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/12156624/1ee8bbca-50f7-479b-9f88-c11d942529dd

#### Uploading a PDF doc
https://github.com/Expensify/App/assets/12156624/33d2dad4-34ae-4792-8b1c-69797e1e54b9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/3643eb17-3455-4d9b-882e-ac3fd3242d45


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12156624/7662c9fa-3380-4d07-b77a-8bce1447df6a


https://github.com/Expensify/App/assets/12156624/4eaade54-cedb-4386-9943-641e355d78e0


https://github.com/Expensify/App/assets/12156624/1ae62a27-14a7-40f0-ae9e-ec5ec3000f22


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/user-attachments/assets/3cc40e08-084a-4a6c-9c04-1d00f401e872

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/2ebd0ba2-6c51-4850-86fd-4051522d7bce

#### Sending while offline
https://github.com/user-attachments/assets/31dc8514-23ba-4461-a16d-bf133413c5f9

#### Sending a zip file
https://github.com/user-attachments/assets/df063f67-37a5-4432-b320-baef69127bf0

#### Sending a pdf doc
https://github.com/user-attachments/assets/7b0f82f0-fcde-4837-a7c3-b4aacb943389


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/12156624/fe69b944-d608-48c6-8101-6bd1e3e1d16a

#### Offline
https://github.com/Expensify/App/assets/12156624/18927050-19b2-47fc-83a4-fedd489b0aca


</details>

kidroca committed:
ReportUtils: simplify optimistic comment constructor logic

- Consolidated the logic for generating `htmlForNewComment` and `textForNewComment`.
- Cleaned up redundant checks for file and text presence.
- Optimized creation of the `reportAction` object.

kidroca committed:
ReportUtils: extract attachment HTML generation to a separate function

- Introduced a new function `getAttachmentHtml` to handle the generation of attachment HTML.
- Updated `addComment` function to use `getAttachmentHtml` for cleaner code.
- Ensured `textForNewComment` is parsed from `htmlForNewComment` for consistency.

kidroca committed:
ReportUtils: rename getAttachmentHtml to getUploadingAttachmentHtml

Renamed the function `getAttachmentHtml` to `getUploadingAttachmentHtml` for better clarity and consistency. The new name more accurately reflects its purpose in generating HTML for uploading attachments.

kidroca committed:
Enhance attachment handling in getUploadingAttachmentHtml function

- Improved `getUploadingAttachmentHtml` function to handle image and video attachments.
- Added conditions to check for MIME types and generate appropriate HTML for image and video sources.
- Updated `htmlForNewComment` to remove an extra newline character.

kidroca committed:
fix: prevent error when custom attachment source attribute is undefined

Added nullish coalescing operator to ensure `attribs[CONST.ATTACHMENT_SOURCE_ATTRIBUTE]` is defined before using `split()`, preventing potential runtime errors.

kidroca committed:
ReportUtils: improve attachment HTML generation logic

- Return HTML string directly within conditionals for image and video attachments.
- Ensure a generic message is returned for other file types.
- Adjust the HTML for new comments to conditionally include line breaks when an attachment is present.
```

kidroca committed:
VideoRenderer: correct video source URL resolution logic

Refactor the order of URL resolution priorities to fix issues with video rendering. Now, the video source URL will first check the ATTACHMENT_SOURCE_ATTRIBUTE, then 'src', and finally 'href'.

kidroca committed:
ReportUtils: improve attachment handling for non-media files

Updated the fallback case for handling non-media files in `src/libs/ReportUtils.ts`. Instead of using a generic uploading message, we now present a link to the file for better user experience.

kidroca committed:
ReportUtils: prevent extra line breaks when comment text is empty

kidroca committed:
ReportUtils: lint fix

kidroca committed:
ReportUtils: fix type errors

kidroca committed:
CONST: add optimistic source attribute for attachments

- Introduced `ATTACHMENT_OPTIMISTIC_SOURCE_ATTRIBUTE` in constants.
- Updated `ReportUtils` to include the new optimistic source attribute in the HTML output for images, videos, and other file types.

kidroca committed:
libs/isReportMessageAttachment: improve attachment message check logic in .ts 🧩

- Extracted attachment regex pattern into a constant `ATTACHMENT_PATTERN`
- Simplified the return condition by leveraging the extracted constant
- Fix function parameter documentation

kidroca committed:
AttachmentCommentFragment: update attachment uploading detection

Updated the `isUploading` condition to use `html.contains(CONST.ATTACHMENT_OPTIMISTIC_SOURCE_ATTRIBUTE)` instead of comparing it to `CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML`, because `CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML` is being removed

kidroca committed:
ContextMenuActions: update attachment detection logic

Since `CONST.ATTACHMENT_UPLOADING_MESSAGE_HTML` is no longer used it's removed from the check

kidroca commented:
As PR the PR Author checklist I'm tagging @Expensify/design to review the UI/UX change

---

@shawnborton 

This PR introduces a UI/UX change with the attachment upload flow
I've brought it up earlier, and you responded here: https://github.com/Expensify/App/issues/42297#issuecomment-2210756684

The difference being introduced is now we render a thumbnail of the attachment instead of the text "Uploading attachment..."

Below you can see some **new** vs **old** samples

#### Attaching an image or video

https://github.com/user-attachments/assets/b61f4a53-71f8-4efe-b879-8518f13449f6

#### Attaching an image while offline

![SBS offline](https://github.com/user-attachments/assets/426be4e8-ac89-4584-af81-7e2d0d108ddd)

#### Attaching a PDF doc

https://github.com/user-attachments/assets/ee5cf151-7a2b-4847-8fbf-2e82df43a722

#### Other files

https://github.com/user-attachments/assets/66d5af92-931a-48ec-8fd4-85c6c1f6be9c

---

The UX is identical on iOS and Android native, there are examples attached inside the Screenshots section in the PR description

kidroca committed:
ContextMenuActions: prettier fix

kidroca committed:
isReportMessageAttachment: Update to pass unit tests

- The message "Uploading Attachment..." no longer exists - we don't need to test for such text to match an attachment action
- All attachments use the `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` it can be used to distinct them
  - Attachments in upload use the same attribute as well

kidroca committed:
Remove `uploading-attachment` custom model from BaseHTMLEngineProvider

The `uploading-attachment` custom model is removed as it is no longer used.

trjExpensify commented:
Going to assign Design for review on this one. Exciting!

dannymcclain commented:
I noticed in some of the videos (I think especially the mobile ones) the thumbnail of the uploaded image jumps/changes size right after submitting it. Is there anything we can do to fix that?

Also, I can't tell exactly what's happening so I could be wrong about this, but why in this video does the one on the left not show the large play icon? It's kinda hard to tell from the video, but it looks like it goes straight into the video controls, which is not something I'm sure we want.
 
![CleanShot 2024-07-15 at 16 09 24@2x](https://github.com/user-attachments/assets/905136c8-ef19-4529-b57d-4167fe69a828)

dubielzyk-expensify commented:
This is looking so goood! The only thing I'd love to remove if possible is this intermediate loading state before it's shown, but I'm not sure of the feasibility of this? 
![CleanShot 2024-07-16 at 08 07 14@2x](https://github.com/user-attachments/assets/4bdeb854-96c3-4236-923c-a339ec5f4f6c)

kidroca committed:
ReportUtils: Enhance data attribute generation

- Enhanced data attribute generation by converting to an array and using `Array.prototype.filter` to remove falsy values.
- Added additional attributes for file width and height.

kidroca commented:
@dannymcclain 
> I noticed in some of the videos (I think especially the mobile ones) the thumbnail of the uploaded image jumps/changes size right after submitting it. Is there anything we can do to fix that?

I've just included a fix about that one - when possible, we'll leverage existing information about width and height and prevent the image from changing size

> Also, I can't tell exactly what's happening so I could be wrong about this, but why in this video does the one on the left not show the large play icon? It's kinda hard to tell from the video, but it looks like it goes straight into the video controls, which is not something I'm sure we want.

Your observation is correct. The video on the left side lacks the large play icon. The video on the right is not actually a video until you press the play icon, it's a thumbnail image generated on the backend and used for preview. If it's desired, we can create a follow-up task to generate such a thumbnail image client-side, or try to alter the logic to display the large play icon over the video (since a thumbnail seems to be available natively anyway)

@dubielzyk-expensify 
> This is looking so goood! The only thing I'd love to remove if possible is this intermediate loading state before it's shown, but I'm not sure of the feasibility of this?

We'll be working on a follow-up to further polish the attachment upload and eliminate such problems

dannymcclain commented:
>Your observation is correct. The video on the left side lacks the large play icon. The video on the right is not actually a video until you press the play icon, it's a thumbnail image generated on the backend and used for preview. If it's desired, we can create a follow-up task to generate such a thumbnail image client-side, or try to alter the logic to display the large play icon over the video (since a thumbnail seems to be available natively anyway)

I think we do want all videos to start as thumbnails with the large play icon. That way the behavior will be consistent across all platforms. And then tapping the thumbnail will trigger the platform-specific behavior (desktop plays in place in the message, mobile launches the attachment modal to view). cc @Expensify/design for a gut check on that.

shawnborton commented:
Totally agree, and to do that, I like this approach:

>  or try to alter the logic to display the large play icon over the video (since a thumbnail seems to be available natively anyway)

So that the play button is always the same size and quality.

shawnborton commented:
This tests pretty well, the only minor thing is that you see a quick flash or blink when the image goes from the optimistic version to the actual uploaded version. Can we do anything to hide that transition?

trjExpensify commented:
> #9402

Yeah, I think we discussed that [here](https://github.com/Expensify/App/issues/9402#issuecomment-2173985034) and agreed to look at the content flash as a follow-up as it's pretty involved.

kidroca commented:
> > This tests pretty well, the only minor thing is that you see a quick flash or blink when the image goes from the optimistic version to the actual uploaded version. Can we do anything to hide that transition?
> 
> Yeah, I think we discussed that [here](https://github.com/Expensify/App/issues/9402#issuecomment-2173985034) and agreed to look at the content flash as a follow-up as it's pretty involved.

The follow-up would not simply address this cosmetic issue, but also allow the client to load the attachment **instantly** from their own device, when possible, which should be 100% of the time for an attachment they've just posted. This would save on network bandwidth, and allow us to eventually apply the same benefit to all (recent) attachments (not only the ones the user uploaded themselves)

It's not as big a change as it sounds, but it would be better to work on it separately in a follow-up

shawnborton commented:
Sounds good - follow up totally works for me.

kidroca committed:
Merge branch 'main' into kidroca/feat/attachment-upload-ux

kidroca commented:
Merged main
Unit tests started to fail, but it's unrelated to my changes. I've reported the problem here: https://expensify.slack.com/archives/C01GTK53T8Q/p1721375022329119

kidroca commented:
> BUG: Popup shows the file name instead of video Preview.
> 
> Bug: Preview gets closed automatically and chat is redirected to some other chat. I think it happens when optimistic chat is replaced.
> 
>  20.07.2024_15.05.01_REC.mp4

Maybe there was a mixup with the video upload, as it doesn't seem related to the above bugs: https://github.com/Expensify/App/pull/44889#issuecomment-2241153042

#### BUG: Popup shows the file name instead of video Preview.
Perhaps it's a video that can be played only after processing on the backend. 
Does the video file type change if you download the same attachment from the backend?

---

#### Bug: Preview gets closed automatically and chat is redirected to some other chat

Opening an attachment while it's still uploading indeed causes the attachment viewer to close when the address is replaced with the server value. This would be addressed by the follow-up we discussed earlier as it would cache file locations using onyx and resolve them locally when possible. An alternative might be to not close the Attachment Carousel instantly when an image is removed, and continue to see the image inside the carousel until the user decides to close it themselves 

Being redirected to a different chat doesn't seem related to the changes in this PR. Maybe it's a side effect that happens when the Attachment Viewer is closed automatically 

---

> BUG: While online, clicking on download on a text file, result in errored tab. (This happens with optimistic file until it is replaced with server response).
> 
> Steps:
> 
> 1. Send a text file.
> 2. Immediately click download on it.

Can someone recommend what should we do here? I think I can make changes to let users download the very same file they are still uploading, but it seems unnecessary. It would add special case logic for a very uncommon flow. Perhaps we can hide the download button while file is still being uploaded?

---

> BUG: I see this while sending a video offline. No Preview.

This is not a bug from the current PR, it's how sending a video already works on prod/staging. When offline, trying to send a video results in "You appear to be offline screen"

---

> BUG: No preview for video while offline. 

I'm not sure if this is the same as the one above. The inline preview of the video works while offline, but when you open it in the attachment carousel you see the offline message. I guess that unlike images, no-one anticipated that videos might be played while offline when the video feature was added to the carousel. Same as images who work offline via cache, videos should be able to be previewed if available via cache - this should be handled separately as it's an existing issue

---

> BUG: No preview for PDF while offline.
> BUG: No preview for PDfs until the message is replaced with server response.

The backend generates a thumbnail image for document preview, this is why for documents we display a generic download link until they are uploaded and replaced by the backend. I've described this as _"Verify that a **download link** for the PDF appears immediately in the chat."_, but I'll update the test instructions to mention that this link should be replaced with a doc preview thumbnail once the upload is complete
 
---

### Here's what actionable by me right now

1. **No preview for PDF while offline**
2. **No preview for PDFs until the message is replaced with server response**
   - Action: Update the test steps to reflect the expected behaviour.

### And what's actionable after some feedback

3. **Clicking on download on a text file while online results in an errored tab**
   - Awaiting recommendations on the path forward to address this bug.

4. **Popup shows the file name instead of video Preview**
   - Awaiting more details or the correct video example to proceed.

### Pre-existing Issues for a Follow-up PR

1. **Preview gets closed automatically and chat is redirected to some other chat**
   - Being redirected to a different chat is not related to the changes in this PR and might be a side effect that needs further investigation.

2. **I see this while sending a video offline. No Preview**
3. **No preview for video while offline**
   - These issues are pre-existing and should be handled separately.

trjExpensify commented:
>Can someone recommend what should we do here? I think I can make changes to let users download the very same file they are still uploading, but it seems unnecessary. It would add special case logic for a very uncommon flow. Perhaps we can hide the download button while file is still being uploaded?
>
>cc: @deetergp @trjExpensify What do you think?

Hiding the download button until the file has uploaded makes sense to me. That really isn't an action you take immediately while uploading a file.

deetergp commented:
Agreed! It seems a bit silly to show the download button while the file is uploading.

kidroca committed:
Merge branch 'main' into kidroca/feat/attachment-upload-ux

kidroca committed:
🐛 fix(AnchorForAttachmentsOnly): update download icon visibility logic

Modified the condition for displaying the download icon in `BaseAnchorForAttachmentsOnly.tsx`.
The icon will now only show if the `sourceID` is provided and the application is not offline.
This change ensures that the download button does not appear while the file is still uploading and lacks a final hosted URL

kidroca commented:
Made the change to hide the download button while the file is still being uploaded, here's how it looks like:

#### Sending a zip file
https://github.com/user-attachments/assets/df063f67-37a5-4432-b320-baef69127bf0

#### Sending a pdf doc
https://github.com/user-attachments/assets/7b0f82f0-fcde-4837-a7c3-b4aacb943389

---

✅ Synced `main` and resolved conflicts
✅ Ready for Review

kidroca commented:
Typescript Checks seem to be failing because of something inside `src/libs/actions/Report.ts` which haven't been changed in this PR

kidroca committed:
🐛 fix(components/AnchorForAttachmentsOnly): add sourceID check to download logic

Update the download logic to include a check for the presence of
sourceID alongside existing conditions for isDownloading and
isOffline, preventing potential errors when attempting to download
attachments without a valid source ID.

kidroca commented:
> BUG: Failed download issue also happens to PDF attachments. There is no download button until the preview is shown but clicking the whole name triggers download action.

Added a small fix

parasharrajat reviewed (changes requested):
A bug above. 

We should create issues to track the future or follow-up items if we are not doing those in this PR.

kidroca committed:
♻️ refactor(components/Attachments): improve attachment file name extraction

Update the logic for extracting attachment file names by using the original
file name attribute if available. This enhances clarity and ensures that
the most accurate file name is utilized, falling back to the source only
when necessary.

kidroca committed:
🐛 fix(components/AttachmentModal): prevent downloading local file sources

Added a check to ensure local file sources (file: or blob:) are not
included in the download options. This avoids redundancy and potential
issues with file not found when attempting to download local files.

kidroca commented:
> BUG: Downloading the video file from preview downloads a text file.

I've added another small patch to resolve the following issues:
- ✅ BUG: Downloading the video file from preview downloads a text file.
- ✅ BUG: Popup shows the file name instead of video Preview.

https://github.com/user-attachments/assets/a84ba42d-98da-4b80-a368-1ac7c2f2f1f9

--- 

> We should create issues to track the future or follow-up items if we are not doing those in this PR.

Can we just track them under the existing ticket? I can list them in a status update.

kidroca committed:
🐛 fix(report/ContextMenu): hide download option for uploading files

Added a check to ensure the download option is hidden for resources that
are currently being uploaded. This prevents users from attempting to
download files that are not yet available, improving the user experience.

kidroca commented:
> Thanks for the changes. But I still see following issues.
> 
> 1. Looks like the name of the video file is clipped. rename the test video to `29.07.2024_20.26.50_REC.mp4` before testing.
> 2. Context Menu download still have issues when we try to download a optimistic file.
> 3. Downloaded video is document type. it does not have any extension in the name.

Removed the download option from the context menu for pending uploads. 
It shouldn't be possible to download the video while it's being uploaded.
After upload the file downloads with the correct type and extension. Original name isn't preserved

kidroca committed:
🐛 fix(libs/ReportUtils): revert translationKey assignment for attachments

Restores original logic to set translationKey only when the comment is an attachment

kidroca commented:
> I am fine with not solving the clipped name issue in this PR. But downloaded file should return the extension for correct type.

The clipped filename was happening only if you try to download a file that is not yet uploaded to the backend, no? It's no longer possible to download before the upload completes, so there shouldn't be any issues with the filename and extension.
Once the video is uploaded to the backend, it receives a unique name, and there are no problems downloading it.

parasharrajat reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

:ribbon: :eyes: :ribbon:  C+ reviewed

kidroca commented:
> Future Actionable items are mentioned in [#44889 (comment)](https://github.com/Expensify/App/pull/44889#issuecomment-2247798213). @kidroca @deetergp Let's create a plan for these so we don't see a deploy blocker popping up after merge.

I've just posted my plan for these in the parent issue: https://github.com/Expensify/App/issues/9402#issuecomment-2258312655

deetergp commented:
LGTM

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/10186016478) to staging by https://github.com/deetergp in version: 9.0.15-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/10208759214) to production by https://github.com/roryabraham in version: 9.0.15-9 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

hungvu193 commented on src/components/AnchorForAttachmentsOnly/BaseAnchorForAttachmentsOnly.tsx:
```diff
@@ -63,7 +63,7 @@ function BaseAnchorForAttachmentsOnly({style, source = '', displayName = '', dow
                     <AttachmentView
                         source={sourceURLWithAuth}
```
Since we introduced local file for preview attachments, we also need to check here whether we should not add an encrypted token for the local file. 
More details: https://github.com/Expensify/App/issues/54640#issuecomment-2564648498
